### PR TITLE
fix: Flicker b/w Dark and Light theme

### LINKED ISSUE #15922: [CAL-4095] Embedded Calendar View Flickering between Light and Dark mode (bug)
### Issue Summary

Hi there! I have an embedded [cal.com](http://cal.com) link in my website for people to schedule calls with. My website has a dark theme so I set the javascript to initialize the theme in dark mode:

```javascript
Cal.ns['30min']('ui', {
          theme: 'dark', // set theme to dark mode
          styles: { branding: { brandColor: '#3aaf4b' } },
          hideEventTypeDetails: false,
          layout: 'month_view',
        });
```

Occasionally when I load the page, the calendar will flicker between light and dark themes infinitely. It looks like the `class` value in the `<html>` tag keeps switching between `class="light"` and `class="dark"`. Here is a video:

[https://github.com/user-attachments/assets/64f4d89b-26e1-4990-972f-c873dc7ebaaa](https://github.com/user-attachments/assets/64f4d89b-26e1-4990-972f-c873dc7ebaaa)

I did a bit of testing and it appears to be looping between the theme that the javascript value has passed in (`dark`) and the theme of the browser (my browser is `light` theme). Even if they are both the same value (in my case `dark` and `dark`), my browser shows the value continuously being re-rendered. Watch the video below that shows the `<html>` objects class being continuously updated.

[https://github.com/user-attachments/assets/65d0cafc-eeac-4cf8-b2c7-22d8a6f26041](https://github.com/user-attachments/assets/65d0cafc-eeac-4cf8-b2c7-22d8a6f26041)

### Steps to Reproduce

Embed a [cal.com](http://cal.com) month view in your website using the instructions in the web app and refresh the page enough times.

### Expected behavior

The calendar should not flicker.

### Other information

*No response*

### Screenshots

*No response*

### Environment

- [ ] [app.cal.com](http://app.cal.com)
- [X] Self-hosted Cal

### Desktop (please complete the following information)

```markdown
- Devices: Macbook Pro (16inch) running Mac OS Sonoma and a M1 Macbook Pro
- Browsers: Multiple Chromium browsers including Google Chrome and Microsoft Edge
- The calendar was embedded using the vanilla JS script provided by cal.com's website.



      <div
        style="width: 100%; height: 100%; overflow: scroll"
        id="my-cal-inline"
      ></div>
      <script type="text/javascript">
        (function (C, A, L) {
...
```

```
```

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-4095](https://linear.app/calcom/issue/CAL-4095/embedded-calendar-view-flickering-between-light-and-dark-mode-bug)</sub>

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

- Fixes #15922 
- Fixes CAL-4095

It is fixed by passing `theme` through config which makes theme available to iframe before it even starts loading. It ensures that theme is always available and then it allows us to correctly calculate the storageKey for next-theme(as per the logic here https://github.com/calcom/cal.com/blob/b1350dfd5916f1ad06ec62f15818755aa64cb8f9/apps/web/lib/app-providers.tsx#L235)

Note that the actual fix is just ensuring that the Embed Code passes config.theme. This PR is just ensuring that it reflects in all types of codes.

## Root Cause:
I couldn't identify any change done recently that caused this issue. The only possible explanation is that somehow
`CalcomThemeProvider` is rendering before the parent sends the UI instruction to iframe. It could happen for one of two reasons
- Somehow Next.js is rendering it faster
- Somehow browser is delaying the postmessages

Whatever be the case, there was no guarantee in the earlier approach that it will always work but now it is guaranteed by passing through query param.

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
It is difficult to replicate the bug. So, we need to just ensure that the theme is passed via query parameter of iframe.
image

### CODE REVIEW

**Comment by hariombalhara** [comment]
* **#16042** <a href="https://app.graphite.dev/github/pr/calcom/cal.com/16042?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hariombalhara and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/features/embed/Embed.tsx [hunk 1]
```diff
@@ -575,15 +575,24 @@ const EmbedTypeCodeAndPreviewDialogContent = ({
...
-    layout: BookerLayouts.MONTH_VIEW,
-    floatingPopup: {},
-    elementClick: {},
+    layout: defaultLayout,
+    floatingPopup: {
+      config: defaultConfig,
+    },
+    elementClick: {
+      config: defaultConfig,
+    },
```
Comment (hariombalhara) [inline packages/features/embed/Embed.tsx:595]: It makes all types of embed consistent and supports passing config

File: packages/features/embed/Embed.tsx [hunk 2]
```diff
@@ -707,11 +716,11 @@ const EmbedTypeCodeAndPreviewDialogContent = ({
 
   const FloatingPopupPositionOptions = [
     {
-      value: "bottom-right",
+      value: "bottom-right" as const,
```
Comment (hariombalhara) [inline packages/features/embed/Embed.tsx:719]: Allows better type-checking by ensuring not just any string is acceptable.

File: packages/features/embed/lib/EmbedCodes.tsx [hunk 3]
```diff
@@ -118,29 +127,39 @@ export const Codes = {
       uiInstructionCode: string;
       previewState: PreviewState;
       namespace: string;
+      theme: string;
     }) => {
       return code`${getApiNameForVanillaJsSnippet({ namespace, mainApiName: "Cal" })}("inline", {
-	elementOrSelector:"#my-cal-inline",
-	calLink: "${calLink}",
-	layout: "${previewState.layout}"
```
Comment (hariombalhara) [inline packages/features/embed/lib/EmbedCodes.tsx:125]: `layout` was supposed to be inside config. This change automatically fixes the bug.

File: packages/features/embed/lib/EmbedCodes.tsx [hunk 4]
```diff
@@ -44,22 +44,31 @@ export const Codes = {
 	return <Cal ${namespaceProp}
 	  calLink="${calLink}"
 	  style={{width:"${width}",height:"${height}",overflow:"scroll"}}
-	  ${previewState.layout ? `config={{layout: '${previewState.layout}'}}` : ""}
+	  config={${JSON.stringify(previewState.config)}}
```
Comment (hariombalhara) [inline packages/features/embed/lib/EmbedCodes.tsx:47]: No need to pick the props now. Just use config directly.

File: packages/features/embed/types/index.d.ts [hunk 5]
```diff
@@ -1,20 +1,40 @@
...
-    height: string;
-  };
+  inline: Brand<
+    {
+      width: string;
+      height: string;
+      config?: EmbedConfig;
+    },
+    "inline"
+  >;
```
Comment (hariombalhara) [inline packages/features/embed/types/index.d.ts:20]: This guarantees that we don't accidentally mix one embedType's data with other as they might be structurally similar.

File: packages/features/embed/lib/EmbedTabs.tsx [hunk 6]
```diff
@@ -199,32 +199,28 @@ const getEmbedTypeSpecificString = ({
   if (!frameworkCodes[embedType]) {
     throw new Error(`Code not available for framework:${embedFramework} and embedType:${embedType}`);
   }
+
+  const codeGeneratorInput = {
+    calLink,
+    uiInstructionCode: getEmbedUIInstructionString(uiInstructionStringArg),
+    embedCalOrigin,
+    namespace,
+  };
```
Comment (hariombalhara) [inline packages/features/embed/lib/EmbedTabs.tsx:208]: Now, we have ensured that all methods accept the same structure of data. So, it is possible to use a common object.

File: packages/features/embed/lib/EmbedTabs.tsx [hunk 7]
```diff
@@ -199,32 +199,28 @@ const getEmbedTypeSpecificString = ({
...
   } else if (embedType === "floating-popup") {
-    const floatingButtonArg = {
-      calLink,
-      ...(doWeNeedCalOriginProp(embedCalOrigin) ? { calOrigin: embedCalOrigin } : null),
-      ...previewState.floatingPopup,
-    };
     return frameworkCodes[embedType]({
-      namespace,
-      floatingButtonArg: JSON.stringify(floatingButtonArg),
-      uiInstructionCode: getEmbedUIInstructionString(uiInstructionStringArg),
```
Comment (hariombalhara) [inline packages/features/embed/lib/EmbedTabs.tsx:219]: Taken care of by the code function itself now, similar to others.

**Review by zomars — APPROVED** [review]
LGTM

**Pull request merged**